cmd/libsnap-confine-private/cgroup-freezer-support.c

```diff
@@ -65,3 +65,84 @@ void sc_cgroup_freezer_join(const char *snap_name, pid_t pid)
 	debug("moved process %ld to freezer cgroup hierarchy for snap %s",
 	      (long)pid, snap_name);
 }
+
+bool sc_cgroup_freezer_occupied(const char *snap_name, uid_t uid)
+{
+	// Format the name of the cgroup hierarchy.
+	char buf[PATH_MAX] = { 0 };
+	sc_must_snprintf(buf, sizeof buf, "snap.%s", snap_name);
+
+	// Open the freezer cgroup directory.
+	int cgroup_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	cgroup_fd = open(freezer_cgroup_dir,
+			 O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+	if (cgroup_fd < 0) {
+		die("cannot open freezer cgroup (%s)", freezer_cgroup_dir);
+	}
+	// Open the proc directory.
+	int proc_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	proc_fd = open("/proc", O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+	if (proc_fd < 0) {
+		die("cannot open /proc");
+	}
+	// Open the hierarchy directory for the given snap.
+	int hierarchy_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	hierarchy_fd = openat(cgroup_fd, buf,
+			      O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+	if (hierarchy_fd < 0) {
+		if (errno == ENOENT) {
+			return false;
+		}
+		die("cannot open freezer cgroup hierarchy for snap %s",
+		    snap_name);
+	}
+	// Open the "cgroup.procs" file. Alternatively we could open the "tasks"
+	// file and see per-thread data but we don't need that.
+	int cgroup_procs_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	cgroup_procs_fd = openat(hierarchy_fd, "cgroup.procs",
+				 O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+	if (cgroup_procs_fd < 0) {
+		die("cannot open cgroup.procs file for freezer cgroup hierarchy for snap %s", snap_name);
+	}
+
+	FILE *cgroup_procs SC_CLEANUP(sc_cleanup_file) = NULL;
+	cgroup_procs = fdopen(cgroup_procs_fd, "r");
+	if (cgroup_procs == NULL) {
+		die("cannot convert tasks file descriptor to FILE");
+	}
+	cgroup_procs_fd = -1;	// tasks_fd will now be closed by fclose.
+
+	char *line_buf SC_CLEANUP(sc_cleanup_string) = NULL;
+	size_t line_buf_size = 0;
+	ssize_t num_read;
+	struct stat statbuf;
+	do {
+		num_read = getline(&line_buf, &line_buf_size, cgroup_procs);
+		if (num_read < 0 && errno != 0) {
+			die("cannot read next PID belonging to snap %s",
+			    snap_name);
+		}
+		if (num_read <= 0) {
+			break;
+		}
+		if (num_read > 0 && line_buf[num_read - 1] == '\n') {
+			line_buf[num_read - 1] = '\0';
+		}
+		debug("found process id: %s\n", line_buf);
+
+		if (fstatat(proc_fd, line_buf, &statbuf, AT_SYMLINK_NOFOLLOW) <
+		    0) {
+			// The process may have died already.
+			if (errno != ENOENT) {
+				die("cannot stat /proc/%s", line_buf);
+			}
+		}
+		if (statbuf.st_uid == uid || uid == (uid_t) - 1) {
+			debug("found process %s belonging to user %d",
+			      line_buf, statbuf.st_uid);
+			return true;
```

Comment:
It isn't clear why you are checking the uid here? Not to mention, the debug statement is a white lie when `uid == (uid_t) - 1`. Because we don't (yet) have per-user mounts, if we call this function with something other than `(uid_t) - 1`, then this function will say the cgroup is unoccupied, but it is occupied.

Reply:
The uid check was a nod towards per-user mounts where we wanted to freeze a subset of UIDs.

Reply:
I thought that might've been the case. I think I'd prefer to see a 'TODO' comment here saying we may also consider select uids when we have per-user mounts, then change the logic accordingly.

Reply:
Sure, I'll do that